File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
# nums = [1, -1, 0] k = 0
# prefixSum = [0, 1, 0, 0]

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        cnt = 0
        
        hashmap = {}
        hashmap[0] = 1
        
        prefixSum = [0]
        for val in nums:
            prefixSum.append(prefixSum[-1] + val)
        
        for i in range(1, len(prefixSum)):
            if prefixSum[i] - k in hashmap:
                cnt += hashmap[prefixSum[i] - k]
            hashmap[prefixSum[i]] = hashmap[prefixSum[i]] + 1 if(prefixSum[i]in hashmap) else 1
        return cnt
            

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int[] prefixSum = new int[nums.length];
        prefixSum[0] = nums[0];
        for(int i = 1; i<nums.length; i++){
            prefixSum[i] = prefixSum[i-1] + nums[i];
        }
        HashMap<Integer, Integer> mp = new HashMap<>();
        mp.put(0,1);
        int ans = 0;
        for(int val : prefixSum){
            int remain = val - k;
            if(mp.containsKey(remain)){
                ans += mp.get(remain);
            }
            if(mp.containsKey(val)){
                mp.put(val, mp.get(val)+1);
            }else{
                mp.put(val, 1);
            }
        }
        return ans;
    }
}